### DISCUSSION TITLE
Improving the efficiency of steiner_tree (kou vs. mehlhorn)

### DISCUSSION THREAD

**Discussion by aartinian** [networkx/networkx#7613] [Q&A]
Hi everyone,

I’m currently working on a project involving district heating networks (radial, no loops), specifically focusing on scenarios with a single energy source. To optimize the network's layout, I’ve been experimenting with `steiner_tree` and have been switching between the two available methods: `kou` and `mehlhorn`.

Here’s what I’ve observed:

• `kou`: delivers significantly better results in terms of the optimality of the network, but it has some downsides. It consumes an excessive amount of memory and tends to be quite slow, especially as the network size grows.

• `mehlhorn`: on the other hand, this method is incredibly fast and uses minimal memory, but the results are sometimes noticeably suboptimal compared to `kou`.

To address these issues, I tried optimizing `kou` by reducing the scope of the metric closure determination to only include relevant terminal elements rather than all nodes in the graph. This improvement has helped to some extent, but I’m still looking for a way to achieve the quality of results that `kou` provides, with the speed and efficiency of `mehlhorn`.

I'm looking for suggestions or insights on the following:

1. Understanding the differences: What are the core differences between the `kou` and `mehlhorn` algorithms in terms of their underlying processes? Why does `kou` consume more resources, and what makes `mehlhorn` so much faster but less accurate?

2. Further optimization: Are there any known techniques or strategies to optimize `kou` further, potentially borrowing elements from `mehlhorn`, to strike a balance between optimal results and computational efficiency?

3. Best practices: For those who have worked on similar problems, are there best practices or alternative approaches within or outside NetworkX that might help me achieve both the desired performance and accuracy?

Thank you in advance 🙏

**Comment by rossbar** [comment]
Certainly, performance improvements are always a worthwhile consideration!

> Understanding the differences: What are the core differences between the kou and mehlhorn algorithms in terms of their underlying processes? Why does kou consume more resources, and what makes mehlhorn so much faster but less accurate?

AIUI these are two completely separate algorithms for solving the Steiner tree problem. I'm no expert, but the first place I'd look to get a fuller understanding of the two algorithms are the [original papers referenced in the docstring](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.approximation.steinertree.steiner_tree.html).

Now, re: the specifics of how these algorithms are implemented in NetworkX - that's another thing that would require some legwork to digest and understand. It is certainly possible that the implementation can be improved (e.g. with respect to memory footprint or computation time), but it will take some investigation. Perhaps line profiling would be helpful.

> Further optimization: Are there any known techniques or strategies to optimize kou further, potentially borrowing elements from mehlhorn, to strike a balance between optimal results and computational efficiency?

Aside from looking for more efficient ways to implement the existing algorithms, it's always worth reviewing the literature to see if there are any algorithmic improvements. Personally, I wouldn't approach this problem from the perspective of trying to "borrow" ideas from one algorithm to port to the other --- IMO that's a great way to maximize complexity --- but that's just me!

> Best practices: For those who have worked on similar problems, are there best practices or alternative approaches within or outside NetworkX that might help me achieve both the desired performance and accuracy?

Most of the post above I've written from the perspective of reviewing/improving networkx. Re: solving your immediate blocker, there are two things you could consider:
 1. NetworkX has adopted (and is constantly improving) a [dispatching system](https://networkx.org/documentation/stable/reference/backends.html) to support different computational backends. You may want to look into other computational backends like [nx-cugraph](https://docs.rapids.ai/api/cugraph/nightly/nx_cugraph/nx_cugraph/) or [python-graphblas](https://github.com/python-graphblas/graphblas-algorithms) to see if any backends support `steiner_tree`.
 2. Failing that, you can always consider other graph libraries: [igraph](https://python.igraph.org/en/stable/) and [rustworkx](https://www.rustworkx.org/) are two that come to mind - consider checking what functionality they provide for solving the steiner_tree problem!

**Reply by aartinian** [reply]
Hi @rossbar,

Many thanks for your detailed reply! I really appreciate you taking the time. I will look into your suggestions as they sound very promising. I'll share my findings here in case anyone else has a similar issue to resolve.
